blog-content/the-complete-guide-to-opentelemetry-in-next-js.md

```diff
@@ -0,0 +1,394 @@
+
+
+Today, we'll go through a complete guide to using OpenTelemetry in Python, including the high-level concepts as well as how to send traces and logs to your OpenTelemetry backend of choice.
+
+
+## Setting Up OpenTelemetry for Next.js: Tracing, Logging, and Metrics
+
+Observability is crucial for understanding the performance and behavior of your Next.js application. OpenTelemetry (OTel) provides a robust framework for tracing, logging, and metrics collection, allowing you to gain deep insights into how your application operates.
+
+In this guide, we’ll walk through setting up OpenTelemetry in a Next.js project, covering:
+- Tracing: Capturing distributed traces for API requests and page transitions
+- Logging: Collecting structured logs that correlate with traces
+- Metrics: Exporting performance and custom application metrics
+- Built-in Spans: Leveraging Next.js’s automatic spans and debugging with NEXT_OTEL_VERBOSE
+- Exception Tracking: Capturing errors within traces
+- Simplifying Setup with @vercel/otel
+
+By the end of this tutorial, you’ll have OpenTelemetry integrated into your Next.js project with traces, logs, and metrics exported to your preferred backend.
+
+### Installing OpenTelemetry in Next.js
+
+To get started, install the necessary OpenTelemetry dependencies:
+
+```bash
+yarn add @opentelemetry/api @opentelemetry/sdk-node @opentelemetry/instrumentation-http @opentelemetry/instrumentation-fetch @opentelemetry/exporter-trace-otlp-grpc @opentelemetry/exporter-metrics-otlp-grpc @opentelemetry/resources @opentelemetry/semantic-conventions
+```
+
+This setup includes the core OpenTelemetry API, SDK, HTTP and Fetch instrumentations, and OTLP exporters for traces and metrics.
+
+### Setting Up OpenTelemetry SDK
+
+Create a new file `otel.js` at the root of your Next.js project:
+
+```javascript
+import { NodeSDK } from '@opentelemetry/sdk-node';
+import { ConsoleSpanExporter } from '@opentelemetry/sdk-trace-base';
+import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-grpc';
+import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-grpc';
+import { Resource } from '@opentelemetry/resources';
+import { SEMRESOURCENAME } from '@opentelemetry/semantic-conventions';
+import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
+import { FetchInstrumentation } from '@opentelemetry/instrumentation-fetch';
+
+const sdk = new NodeSDK({
+  resource: new Resource({
+    [SEMRESOURCENAME.SERVICE_NAME]: 'nextjs-app',
+  }),
+  traceExporter: new OTLPTraceExporter(), 
+  metricExporter: new OTLPMetricExporter(), 
+  instrumentations: [new HttpInstrumentation(), new FetchInstrumentation()],
+});
+
+sdk.start();
+console.log('OpenTelemetry initialized');
```

Comment:
Think we should mention anything about this not being compatible with edge runtime and tease the future state we'll get to that supports it later on?